### PR TITLE
feat(driver): add blob datasource

### PR DESCRIPTION
**Pull request by YoGhurt111**

### CODE REVIEW

File: driver/txlist_fetcher/blob_datasource.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	BlobHash      string `json:"blob_hash"`
+	KzgCommitment string `json:"kzg_commitment"`
+	Blob          string `json:"blob"`
+}
+
+type BlobDataSeq struct {
+	Data []*BlobData `json:"data"`
+}
+
+func NewBlobDataSource(
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:32]: Why not move this into pkg/rpc module?

File: driver/config.go [hunk 2]
```diff
@@ -44,6 +46,15 @@ func NewConfigFromCliContext(c *cli.Context) (*Config, error) {
 		return nil, errors.New("empty L1 beacon endpoint")
 	}
 
+	var blobServerEndpoint *url.URL
```
Comment (mask-pp) [inline driver/config.go:49]: Just using its string type is enough?

File: driver/txlist_fetcher/blob_datasource.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+				Blob:          value.Blob,
+			}
+		}
+	}
+	err = nil
+	return sidecars, err
+}
+
+// getBlobFromServer get blob data from server path `/getBlob`.
+func (ds *BlobDataSource) getBlobFromServer(ctx context.Context, blobHash string) (*BlobDataSeq, error) {
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:85]:
```suggestion
func (ds *BlobDataSource) getBlobFromServer(ctx context.Context, blobHash common.Hash) (*BlobDataSeq, error) {
```

File: driver/txlist_fetcher/blob_datasource.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+	var (
+		route = "/getBlob"
+		param = map[string]string{"blobHash": blobHash}
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:88]:
```suggestion
		param = map[string]string{"blobHash": blobHash.String()}
```

File: driver/txlist_fetcher/blob_datasource.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+	} else {
+		sidecars, err = ds.rpc.L1Beacon.GetBlobs(ctx, meta.Timestamp)
+	}
+	if err != nil {
+		log.Info("Failed to get blobs from beacon, try to use blob server.", "err", err.Error())
+		if ds.blobServerEndpoint == nil {
+			log.Info("No blob server endpoint set")
+			return nil, err
+		}
+		blobs, err := ds.getBlobFromServer(ctx, "0x"+common.Bytes2Hex(meta.BlobHash[:]))
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:68]:
```suggestion
		blobs, err := ds.getBlobFromServer(ctx, meta.BlobHash)
```

File: driver/txlist_fetcher/blob_datasource.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+) *BlobDataSource {
+	return &BlobDataSource{
+		ctx:                ctx,
+		rpc:                rpc,
+		blobServerEndpoint: blobServerEndpoint,
+	}
+}
+
+// GetBlobs get blob sidecar by meta
+func (ds *BlobDataSource) GetBlobs(
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:45]: Simplify the content:
```
if ds.rpc.L1Beacon != nil {

} else if ds.blobServerEndpoint != "" {

} else {
return nil,errBeaconNotFound
}
```

File: driver/txlist_fetcher/blob_datasource.go (hunk 3 above)
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:85]:
```suggestion
func (ds *BlobDataSource) getBlobFromServer(ctx context.Context, blobHash string) ([]*blob.Sidecar, error) {
```

File: driver/txlist_fetcher/blob_datasource.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	"github.com/taikoxyz/taiko-client/pkg/rpc"
+)
+
+type BlobDataSource struct {
+	ctx                context.Context
+	rpc                *rpc.Client
+	blobServerEndpoint *url.URL
+}
+
+type BlobData struct {
```
Comment (mask-pp) [inline driver/txlist_fetcher/blob_datasource.go:22]: Can we choose to use the same struct `blob.Sidecar`?

File: internal/testutils/helper.go [hunk 8]
```diff
@@ -287,7 +287,7 @@ func RandomBytes(size int) (b []byte) {
 func RandomPort() int {
 	port, err := freeport.GetFreePort()
 	if err != nil {
-		log.Crit("Failed to get local free random port", "err", err)
+		log.Crit("Failed to get local free random port", "custom_err", err)
```
Comment (davidtaikocha) [inline internal/testutils/helper.go:290]:
```suggestion
		log.Crit("Failed to get local free random port", "error", err)
```

File: internal/testutils/helper.go [hunk 9]
```diff
@@ -298,7 +298,7 @@ func LocalRandomProverEndpoint() *url.URL {
 
 	proverEndpoint, err := url.Parse(fmt.Sprintf("http://localhost:%v", port))
 	if err != nil {
-		log.Crit("Failed to parse local prover endpoint", "err", err)
+		log.Crit("Failed to parse local prover endpoint", "custom_err", err)
```
Comment (davidtaikocha) [inline internal/testutils/helper.go:301]:
```suggestion
		log.Crit("Failed to parse local prover endpoint", "error", err)
```

File: proposer/proposer_test.go [hunk 10]
```diff
@@ -152,7 +157,7 @@ func (s *ProposerTestSuite) getLatestProposedTxs(
 			case event := <-sink:
 				txs, err := parseTxs(s.RPCClient, event)
 				if err != nil {
-					log.Error("failed to parse txs", "err", err)
+					log.Error("failed to parse txs", "custom_err", err)
```
Comment (davidtaikocha) [inline proposer/proposer_test.go:160]:
```suggestion
					log.Error("failed to parse txs", "error", err)
```

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**